### ISSUE TITLE
Target suffix loading: use of `contains` may pick multiple targets

### ISSUE THREAD

**Issue by uzaymacar** [ivadomed/ivadomed#1055]
### Current behavior

The current implementation of target suffix loading involves the use of the `contains()` string method to capture target files as shown in the below lines:

https://github.com/ivadomed/ivadomed/blob/76b36a0a0f7141feb2d5b00b33e4c3a06865fc2c/ivadomed/loader/bids_dataframe.py#L125-L128

This means that

* if your derivatives contains multiple filenames with overlap (e.g. `_lesion-manual.nii.gz` and `_lesion-manual2.nii.gz`) and
* if  you use a `"target_suffix"` containing the overlap (e.g. `_lesion-manual`),

then the current loading process picks all of the filenames with the given overlap as potential targets. 

This is problematic because the user might (reasonably) think that only `_lesion-manual.nii.gz` will be used as the ground-truth when the `"target_suffix"` is specified as `_lesion-manual`, whereas in reality both `_lesion-manual.nii.gz` and `_lesion-manual2.nii.gz` are used. 

I didn't go deeper in the codebase yet to see what happens when multiple targets are picked during the loading process, but I can confirm that the GT that ends up being used changes from run to run based on my experiments with `ivadomed --test`.

### Expected behavior
 

I would expect the `"target_suffix"` to **exactly match** the target filename. Two options we have are:
1. Changing line 128 as shown above with something like
   ```
   & df_next['filename'].str.split(os.extsep).apply(lambda x: x[0]).str.endswith('|'.join(self.target_suffix)))]
   ```
   which enforces an exact match of target suffix.

2. In case the use of `contains()` is justified in some scenarios and its implemented that way for a specific reason, the user can instead give a full `"target_suffix"` including the file extension such as `_lesion-manual.nii.gz`. I have tested this and can confirm it solves the problem.

The latter option is notably a lot easier to implement as it doesn't require a change in the codebase. However, I think this is an issue many people might face without knowing it in the future.

### Steps to reproduce

1. Download the `basel-mp2rage` dataset as shown [here](https://github.com/ivadomed/model_seg_ms_mp2rage#get-the-data).
2. Run preprocessing on the data as shown [here](https://github.com/ivadomed/model_seg_ms_mp2rage#prepare-the-data).
3. Insert `print(bids_df.get_deriv_fnames())` [after line 370 in `ivadomed/main.py`](https://github.com/ivadomed/ivadomed/blob/76b36a0a0f7141feb2d5b00b33e4c3a06865fc2c/ivadomed/main.py#L370) which initializes the BIDS data frame.
4. Run the [following config file](https://github.com/ivadomed/model_seg_ms_mp2rage/blob/main/config/seg_lesion.json) with `ivadomed --train`.
5. Take a look at the output of the print statement from step 3.

However, this is specific to one dataset and it is understandable that not everybody will be willing to run preprocessing etc. on this particular dataset. In this case, I would like to point out that this issue is reproducible with **any dataset which has multiple annotations per subject where the string target suffix for these annotations have an overlap**.

**Comment by mariehbourget** [comment]
Thank you @uzaymacar for reporting this issue!

I don't see any specific cases where `contains()` is beneficial, I agree with you that it should look for an **exact match** to avoid the "silent" issue you ran into.

I would be in favor of option 1 as it prevents such "silent" issue.
Option 2 would work as well but not all derivatives files have an `nii.gz` extension (for MRI it could be `.nii` as well, and for microscopy `.png` and `.tif`)

**Comment by uzaymacar** [comment]
Thanks for your input and insights @mariehbourget! I will bring this up in our next internal meeting, but two things I would like to think more are: 
* there might be previous config files that rely on the `contains()` implementation similar to what is shown [here](https://github.com/ivadomed/model_seg_ms_mp2rage/issues/43#issue-1114565060).
* my suggestion above with `endswith()` instead of `contains()` is *still* not an exact match as technically `manual` could be used to match `lesion-manual`

**Comment by mariehbourget** [comment]
> * there might be previous config files that rely on the `contains()` implementation similar to what is shown [here](https://github.com/ivadomed/model_seg_ms_mp2rage/issues/43#issue-1114565060)

I'm not sure to understand and may be missing some context.
So you would have several GT with the name `_seg-*` (ex: `_seg-01`, `_seg-02`, `_seg-03`) and put` _seg` in the config file.
Then either one of those GT would be paired with the image for training?

I think right now the loader will associate only 1 GT (ex: `_seg-01`) to 1 given image. But I'm not sure if you re-run the training, if it will always take `_seg-01` or would take another one ( ex: `_seg-02`) depending on the order of the file indexing?
Isn't that something we would want to avoid like in your issue https://github.com/ivadomed/model_seg_ms_mp2rage/issues/41?

> * my suggestion above with `endswith()` instead of `contains()` is _still_ not an exact match as technically `manual` could be used to match `lesion-manual`

In that case, maybe we could split the filename after the `modality_suffix` (`T1w`, `SEM`, etc) which is the column `suffix` in the `bids_dataframe`, remove the extension and look for an exact match (instead of endswith())?

Those are very interesting questions, I'll think about it!

**Comment by uzaymacar** [comment]
> I'm not sure to understand and may be missing some context.

The first point of mine simply says that people might have used, for example, `_seg` in their config file to load `_seg-XX` as the GT. This might even be the case for datasets with a single GT. As a result of the PR that will address this issue, their config files will no longer work (as we want exact match).

> But I'm not sure if you re-run the training, if it will always take _seg-01 or would take another one ( ex: _seg-02) depending on the order of the file indexing?

Indeed! This is what I have observed: it takes another one and which one changes from run to run (observed with `ivadomed --test`). And yes, the goal of this issue is to avoid this.

> maybe we could split the filename after the modality_suffix

This is definitely the next best idea I was looking for 🙂.

**Comment by mariehbourget** [comment]
While debugging for another issue #1096 from an external user, I came across this exact problem.
In this case, we have the following target_suffix (coming from the segmentation in ADS):
`_seg-axon`
`_seg-axonmyelin`
`_seg-myelin`
With `_seg-axon` in the config file, both `_seg-axon` and `_seg-axonmyelin` are picked up which is not good at all.

I'll inform the user for now and tag this issue as high priority.

**Comment by mariehbourget** [comment]
Hi @uzaymacar, did you already started working on this issue?
I would like to put it in priority but don't want us to work in parallel and duplicate our efforts.

**Comment by uzaymacar** [comment]
Feel free to take it on @mariehbourget! I only created a branch so it's OK.

**Comment by mariehbourget** [comment]
> > maybe we could split the filename after the modality_suffix
> 
> This is definitely the next best idea I was looking for 🙂.

Unfortunately, my previous idea is not working, because the filename entities, including the modality suffix are not indexed in the dataframe for derivatives.
